core/src/nozzle_subgraph/monitor.rs

```diff
@@ -0,0 +1,573 @@
+                debug!(logger, "Removing active subgraph");
+                subgraph_instances.remove(&deployment);
+            }
+            Some(_subgraph_instance) => {
+                debug!(logger, "Active subgraph does not need to be removed");
+                return;
+            }
+            None => {
+                debug!(logger, "Subgraph is not active");
+            }
+        }
+
+        if let Some(pending_start_command) = pending_start_commands.remove(&deployment) {
+            debug!(logger, "Resending a pending start command");
+            handle_send_result(&logger, command_tx.send(pending_start_command));
+        }
+    }
+
+    /// Spawns a background process that executes the subgraph runner future.
+    ///
+    /// An additional background process is spawned to handle the graceful shutdown of the subgraph runner,
+    /// and to ensure correct behaviour even if the subgraph runner panics.
+    fn start_subgraph(
+        logger: Logger,
+        cancel_token: CancellationToken,
+        id: u32,
+        deployment: DeploymentLocator,
+        runner: BoxRunner,
+        command_tx: mpsc::UnboundedSender<Command>,
+    ) -> SubgraphInstance {
+        let mut runner_handle = tokio::spawn({
+            let logger = logger.new(slog::o!("process" => "subgraph_runner"));
+            let cancel_token = cancel_token.cheap_clone();
+
+            async move {
+                info!(logger, "Subgraph started");
+
+                match runner(cancel_token).await {
+                    Ok(()) => {
+                        info!(logger, "Subgraph stopped");
+                    }
+                    Err(e) => {
+                        error!(logger, "Subgraph failed";
+                            "error" => ?e
+                        );
+                    }
+                }
+            }
+        });
+
+        let supervisor_handle = tokio::spawn({
+            let logger = logger.new(slog::o!("process" => "subgraph_supervisor"));
+            let cancel_token = cancel_token.cheap_clone();
+
+            fn handle_runner_result(logger: &Logger, result: Result<(), tokio::task::JoinError>) {
+                match result {
+                    Ok(()) => {
+                        debug!(logger, "Subgraph completed execution");
+                    }
+                    Err(e) if e.is_panic() => {
+                        error!(logger, "Subgraph panicked";
+                            "error" => ?e
+                        );
+
+                        // TODO: Maybe abort the entire process on panic and require a full graph-node restart.
+                        //       Q: Should a bug that is triggered in a specific subgraph affect everything?
+                        //       Q: How to make this failure loud enough so it is not missed?
+                        //
+                        // println!("Subgraph panicked");
+                        // std::process::abort();
```

Comment:
We shouldn't do that; yes, it's hard to spot a panic in lots of logs, but taking down the process would mean that one subgraph can affect lots of other subgraphs.